akka-projection-slick/src/main/scala/akka/projection/slick/OffsetStore.scala

```diff
@@ -0,0 +1,91 @@
+class OffsetStore[P <: JdbcProfile](db: P#Backend#Database, profile: P) {
+
+  import profile.api._
+
+  // offset manifest
+  private val StringM = "STR"
+  private val LongM = "LNG"
+  private val IntM = "INT"
+  private val SequenceM = "SEQ"
+  private val TimeBasedUUIDM = "TBU"
+
+  def readOffset[Offset](projectionId: String): Future[Option[Offset]] = {
+    // map using Slick's own EC
+    implicit val ec = db.executor.executionContext
+    val action =
+      offsetTable.filter(_.projectionId === projectionId).result.headOption.map(row => fromRowToOffset[Offset](row))
+
+    db.run(action)
+  }
+
+  def saveOffset[Offset](projectionId: String, offset: Offset): slick.dbio.DBIO[Done] = {
+    // map using Slick's own EC
+    implicit val ec = db.executor.executionContext
+    val (offsetStr, manifest) = stringOffsetAndManifest(offset)
+    offsetTable.insertOrUpdate(OffsetRow(projectionId, offsetStr, manifest)).map(_ => Done)
+  }
+
+  /**
+   * Deserialize an offset from a result row.
+   * The offset is converted from its string representation to its real time.
+   */
+  private def fromRowToOffset[Offset](row: Option[OffsetRow]): Option[Offset] = {
+    row match {
+      case Some(OffsetRow(_, offset, StringM))   => Some(offset.asInstanceOf[Offset])
+      case Some(OffsetRow(_, offset, LongM))     => Some(offset.toLong.asInstanceOf[Offset])
+      case Some(OffsetRow(_, offset, IntM))      => Some(offset.toInt.asInstanceOf[Offset])
+      case Some(OffsetRow(_, offset, SequenceM)) => Some(query.Sequence(offset.toLong).asInstanceOf[Offset])
+      case Some(OffsetRow(_, offset, TimeBasedUUIDM)) =>
+        Some(query.TimeBasedUUID(UUID.fromString(offset)).asInstanceOf[Offset])
+      case Some(OffsetRow(id, offset, manifest)) =>
+        throw new IllegalStateException(s"Unknown offset manifest [$manifest] for projection id [$id]")
+      case None => None
+    }
+  }
+
+  /**
+   * Convert the offset to a tuple (String, String) where the first element is
+   * the String representation of the offset and the second its manifest
+   */
+  private def stringOffsetAndManifest[Offset](offset: Offset): (String, String) = {
+    offset match {
+      case s: String                => s -> StringM
+      case l: Long                  => l.toString -> LongM
+      case i: Int                   => i.toString -> IntM
+      case seq: query.Sequence      => seq.value.toString -> SequenceM
+      case tbu: query.TimeBasedUUID => tbu.value.toString -> TimeBasedUUIDM
+      case _                        => throw new IllegalArgumentException(s"Unsupported offset type, found [${offset.getClass}]")
+    }
+  }
+
+  private class OffsetStoreTable(tag: Tag) extends Table[OffsetRow](tag, "PROJECTION_OFFSET_STORE") {
+
+    def projectionId = column[String]("PROJECTION_ID", O.Length(255, varying = false), O.PrimaryKey)
+    def offset = column[String]("OFFSET", O.Length(255, varying = false))
+    def manifest = column[String]("MANIFEST", O.Length(4))
```

Comment:
@jroper idea of having a status (Ok, Failure) and failure count is interesting. We can explore this idea on a later phase. Preferably before stable version to avoid DB migrations.

Reply:
with a `ext` field?

Reply:
What would be the value stored in the `ext` column?

Reply:
@hepin1989, this will be explored on another PR. I ping you when we start to get your feedback. Thanks.